File: Wisdom_Education_Android/base/src/main/java/com/netease/yunxin/app/wisdom/base/util/PreferenceUtil.java
/*
 * Copyright (c) 2021 NetEase, Inc.  All rights reserved.
 * Use of this source code is governed by a MIT license that can be found in the LICENSE file.
 */

package com.netease.yunxin.app.wisdom.base.util;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.PreferenceManager;

import java.util.UUID;

public class PreferenceUtil {

    private static final String KEY_POLICY_SHOW = "KEY_POLICY_SHOW";

    private static final String KEY_DEVICE_ID = "KEY_DEVICE_ID";

    private static SharedPreferences sharedPreferences;

    public static void init(@NonNull Context context) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(
                context.getApplicationContext());
    }

    public static void put(@NonNull String key, @Nullable Object value) {
        SharedPreferences sp = getSharedPreferences();
        if (value instanceof Boolean) {
            sp.edit().putBoolean(key, (Boolean) value).apply();
        } else if (value instanceof Integer) {
            sp.edit().putInt(key, (Integer) value).apply();
        } else if (value instanceof String) {
            sp.edit().putString(key, (String) value).apply();
        } else if (value instanceof Float) {
            sp.edit().putFloat(key, (Float) value).apply();
        } else if (value instanceof Long) {
            sp.edit().putLong(key, (Long) value).apply();
        }
    }

    public static <T> T get(@NonNull String key, @Nullable T defaultValue) {
        Object result;
        SharedPreferences sp = getSharedPreferences();
        if (defaultValue instanceof Boolean) {
            result = sp.getBoolean(key, (Boolean) defaultValue);
        } else if (defaultValue instanceof Integer) {
            result = sp.getInt(key, (Integer) defaultValue);
        } else if (defaultValue instanceof String) {
            result = sp.getString(key, (String) defaultValue);
        } else if (defaultValue instanceof Float) {
            result = sp.getFloat(key, (Float) defaultValue);
        } else if (defaultValue instanceof Long) {
            result = sp.getLong(key, (Long) defaultValue);
        } else {
            return null;
        }
        return (T) result;
    }

    private static SharedPreferences getSharedPreferences() throws IllegalStateException {
        if (sharedPreferences == null) {
            throw new IllegalStateException("PreferenceManager is not initialized. Please call init() before use!");
        }
        return sharedPreferences;
    }

    public static boolean isShowPolicy() {
        return get(KEY_POLICY_SHOW, false);
    }

    public static void setShowPolicy(boolean isShow) {
        put(KEY_POLICY_SHOW, isShow);
    }

    public static void setDeviceId(String deviceId) {
        put(KEY_DEVICE_ID, deviceId);
    }

    public static String getDeviceId() {
        String deviceId = get(KEY_DEVICE_ID, "");
        if (TextUtils.isEmpty(deviceId)) {
            deviceId = UUID.randomUUID().toString();
            setDeviceId(deviceId);
        }
        return deviceId;
    }
}

File: Wisdom_Education_Android/base/src/main/java/com/netease/yunxin/app/wisdom/base/util/PreferenceUtil.kt
/*
 * Copyright (c) 2021 NetEase, Inc.  All rights reserved.
 * Use of this source code is governed by a MIT license that can be found in the LICENSE file.
 */
package com.netease.yunxin.app.wisdom.base.util

import android.content.Context
import android.content.SharedPreferences
import android.text.TextUtils
import androidx.preference.PreferenceManager
import java.util.*

object PreferenceUtil {
    private const val KEY_POLICY_SHOW = "KEY_POLICY_SHOW"
    private const val KEY_DEVICE_ID = "KEY_DEVICE_ID"
    private const val KEY_REUSE_IM = "KEY_REUSE_IM"
    private const val KEY_RECORD_PLAY_ROOMUUID = "KEY_RECORD_PLAY_ROOMUUID"
    private const val KEY_RECORD_PLAY_RTCID = "KEY_RECORD_PLAY_RTCID"
    private const val KEY_OPEN_CHATROOM = "KEY_OPEN_CHATROOM"
    private const val LOW_LATENCY_LIVE = "LOW_LATENCY_LIVE"
    private var sharedPreferences: SharedPreferences? = null
    fun init(context: Context) {
        if (sharedPreferences != null) return
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(
            context.applicationContext
        )
    }

    private fun put(key: String, value: Any) {
        val sp = getSharedPreferences()
        when (value) {
            is Boolean -> {
                sp!!.edit().putBoolean(key, value).apply()
            }
            is Int -> {
                sp!!.edit().putInt(key, value).apply()
            }
            is String -> {
                sp!!.edit().putString(key, value).apply()
            }
            is Float -> {
                sp!!.edit().putFloat(key, value).apply()
            }
            is Long -> {
                sp!!.edit().putLong(key, value).apply()
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    operator fun <T> get(key: String, defaultValue: T): T {
        val result: Any
        val sp = getSharedPreferences()
        when (defaultValue) {
            is Boolean -> {
                result = sp!!.getBoolean(key, (defaultValue as Boolean))
            }
            is Int -> {
                result = sp!!.getInt(key, (defaultValue as Int))
            }
            is String -> {
                result = sp!!.getString(key, defaultValue as String)!!
            }
            is Float -> {
                result = sp!!.getFloat(key, (defaultValue as Float))
            }
            is Long -> {
                result = sp!!.getLong(key, (defaultValue as Long))
            }
            else -> {
                return defaultValue
            }
        }
        return result as T
    }

    @Throws(IllegalStateException::class)
    private fun getSharedPreferences(): SharedPreferences? {
        checkNotNull(sharedPreferences) { "PreferenceManager is not initialized. Please call init() before use!" }
        return sharedPreferences
    }

    var isShowPolicy: Boolean
        get() = get(KEY_POLICY_SHOW, false)
        set(isShow) {
            put(KEY_POLICY_SHOW, isShow)
        }
    var deviceId: String
        get() {
            var deviceId = get(KEY_DEVICE_ID, "")
            if (TextUtils.isEmpty(deviceId)) {
                deviceId = UUID.randomUUID().toString()
                PreferenceUtil.deviceId = deviceId
            }
            return deviceId
        }
        set(deviceId) {
            put(KEY_DEVICE_ID, deviceId)
        }


    var reuseIM: Boolean
        get() {
            return get(KEY_REUSE_IM, false)
        }
        set(enable) {
            put(KEY_REUSE_IM, enable)
        }

    var recordPlay: Pair<String, String>
        get() {
            val roomUuid = get(KEY_RECORD_PLAY_ROOMUUID, "")
            val rtcCid = get(KEY_RECORD_PLAY_RTCID, "")
            return Pair(roomUuid, rtcCid)
        }
        set(pair) {
            val sp = getSharedPreferences()
            sp!!.edit()
                .putString(KEY_RECORD_PLAY_ROOMUUID, pair.first)
                .putString(KEY_RECORD_PLAY_RTCID, pair.second)
                .apply()
        }

    var enableChatRoom: Boolean
        get() {
            return get(KEY_OPEN_CHATROOM, true)
        }
        set(enable) {
            put(KEY_OPEN_CHATROOM, enable)
        }

    var lowLatencyLive: Boolean
        get() {
            return get(LOW_LATENCY_LIVE, true)
        }
        set(enable) {
            put(LOW_LATENCY_LIVE, enable)
        }

}